### ISSUE TITLE
Formsy Slow on Big Form

### ISSUE THREAD

**Issue by ionutt** [formsy/formsy-react#88]
Hello,
    
  I encounter an problem when using Fomsy with a lot of fields ( over 50 with dropdowns, inputs, date ). It runs realy slow. More than that when I am changing a input realy fast the browser is freezing. 
  Is this a know issue of this library?

Thanks

**Comment by MilosRasic** [comment]
Can't be 100% sure without an example, but it's probably both: a problem with optimization of the components caused by how formsy works. Because it supports validations that depend on values of any other field in the form, formsy will rerender all the fields when one of them changes. If even one of the fields does some demanding processing every time it rerenders, it will affect the performance of the whole form.

The issue I've had was severely unoptimized dropdown component processing it's available options with a .map() even though none of its props changes. This would cause the text inputs in the same form to lag while the user is typing.

If it's in any way possible, I'd comment out the complex components one by one and check the form without them to see which one is causing the issue.

Another way to work around the issue would be to debounce on change on your text fields and textareas.

On the library side, we could look into some optimizations of the HOC so that it doesn't update the wrapped component if nothing changed. I'm not sure how reliably we can tell if any of the field's validations depend on other fields in the form though.

**Comment by ionutt** [comment]
I divided the "form" into multiple forms and now it's working ok. This is the solution for now.

**Comment by msachi** [comment]
I was seeing the same issue with a very large form (five input fields per user x 50 users = ~ 250 fields). It was rendering very slowly (~20 s), sometimes crashed, and the onChange was very sluggish.

I fixed the load / render time by splitting into multiple forms as per @ionutt 's suggestion. However it meant that I had to submit lots of forms with one click, so had to label the forms with `ref`s and then programatically submit them and check for the validation status of each:
```
  this.refs['some-form'].submit()
  return this.refs['some-form'].state.isValid
```

The onChange lag I fixed by using `lodash.debounce` to only fire every 500-1000 ms.

**Comment by munkhorgil** [comment]
Still no improvement for this? Splitting into multiple forms is not a solution.

**Comment by MilosRasic** [comment]
WIP. My state refactoring just passed the first test :smiley_cat: 

No idea if it will turn out more or less performant than the current model, but at least it will make it easier to discover problems and optimze.

**Comment by munkhorgil** [comment]
@MilosRasic Good to know

**Comment by th3fallen** [comment]
Any updates on this?

**Comment by munkhorgil** [comment]
Any update?

**Comment by ThibautMarechal** [comment]
Well. still no update?

**Comment by rkuykendall** [comment]
@munkhorgil @ThibautMarechal @th3fallen With a recent merge of a performance PR into the V2 branch, I believe this is either fixed or greatly improved.

**Comment by ThibautMarechal** [comment]
This is indeed much better. Thanks !

**Comment by rkuykendall** [comment]
@ThibautMarechal Thank you for the feedback! Please don't hesitate to reach out if you have any problems with the 2.0 codebase. It's taking me longer than I hoped to upgrade my own codebase, but I think it's our old awful code's fault, not Formsy's fault. I don't know if I have confidence to do a full 2.x release until I can run it myself or get some testing help from others.

**Comment by felixmosh** [comment]
I struggling with the same issue on the latest version (1.1.5), form with ~100 checkboxes, the browser got stuck for about 10 secs.

From my debugging, looks like it is related to a poor implementation of cloning of the object [here](https://github.com/formsy/formsy-react/blob/v1.x/src/index.js#L58) and [here](https://github.com/formsy/formsy-react/blob/v1.x/src/index.js#L72), it clones exponential amount of fields, not sure what was the reason to do so.

By remove that object.assign all together improved the same form to be rendered under 500 msec.

This is how it looks like
```
 getCurrentValues = () => (
    this.inputs.reduce((data, component) => {
      const { name } = component.props;
      // eslint-disable-next-line no-param-reassign
      data[name] = component.state.value;
      return data;
    }, {})
  )
```

**Comment by rkuykendall** [comment]
Hey @felixmosh, we made a lot of speed improvements to the 2.0 beta branch, which is where all current work is being directed. The breaking changes aren't too extreme so perhaps you can try running that?

The 2.x branch will hopefully be officially released soon.

**Comment by felixmosh** [comment]
@rkuykendall, thanx for the suggestion, and the hard work on the v2!

I will try it, but it is not so simple, cause I'm using `semantic-ui` & `formsy-semantic-ui` as a binding in a huge app, when it will released as stable, I'll push `formsy-semantic-ui` to make the update and then will transition.

BTW, do you know what is the reason to use `object.assign` in the code that I've patched?
It made the form run smoothly.

**Comment by rkuykendall** [comment]
@felixmosh No, it was most likely a mistake which has most likely been fixed. My guess is that, as you can see from this issue, performance was just not a priority in the past. 

`formsy-semantic-ui-react` looks awesome! The author recently merged a PR here so I'm sure it'll be upgraded to Formsy 2.x quickly.

The only thing I'm really waiting on now is getting the Typescript types to be workable, so hopefully very very soon.

**Comment by felixmosh** [comment]
@rkuykendall, cool!
I can PR that fix, will we publish patch version for this fix?

**Comment by rkuykendall** [comment]
@felixmosh You mean like a 1.x patch? I'm not against the notion. We have a 1.x branch that we could use. It's just so ancient, I don't even have a strong grasp of how to release it, since it wasn't automated back then. I think I'd prefer trying to finish the 2.x release.

**Comment by morkeleb** [comment]
I just tried the 2.x version with the @felixmosh's change and it does improve performance.

We're having issues with the 1.x branch.

I tried changing to beta.4 without much performance increase.

Removing the Object.assing copies does however increase the performance to the point where the system becomes workable, still slow but workable.

**Comment by felixmosh** [comment]
I see that v2 has the same `Object.assign` shenanigans, https://github.com/formsy/formsy-react/blob/master/src/index.ts#L191,

There is no need to create a new copy of the object that is reduced, but maybe the value it self.

In my opinion, this code 
```
public getCurrentValues = () =>
  this.inputs.reduce((data, component) => {
    const dataCopy = typeof component.state.value === 'object' ? Object.assign({}, data) : data; // avoid param reassignment
    dataCopy[component.props.name] = component.state.value;
    return dataCopy;
  }, {});
```

is by mistake assigning the reduced value (that is a new instance anyway) instead of copying `component.state.value`.

It should be
```
public getCurrentValues = () =>
  this.inputs.reduce((data, component) => {
    const valueCopy = typeof component.state.value === 'object' ? Object.assign({}, component.state.value) : component.state.value; // avoid param reassignment
    data[component.props.name] = valueCopy;
    return data;
  }, {});
```

```

**Comment by rkuykendall** [comment]
Totally possible that it would help. We have a decent test suite so it should be easy to try out. My focus is definitely on trying to release 2.x but I love performance improvements.
